Arm/refactor/task breakdown

arminsabouri opened the pull request:
And fixed the current liting and fmt warnings we we're getting. Hopefully CI passes now

arminsabouri committed:
refactor(poa): breakout block building and block importing into seperate tasks

arminsabouri committed:
refactor(poa): integrate sync task into block production's `start_task`

arminsabouri committed:
refactor(poa): update `persist_new_block` to use fcu engine utility

arminsabouri committed:
refactor(poa): use engine utils in poa sync controller

arminsabouri committed:
fix(rpc): botanix specific clippy warnings

arminsabouri committed:
fix(poa): clean up poa clippy warnings

arminsabouri committed:
fix(botanix_lib): fix clippy warnings

And snuck some additional clippy fixes found in `src/chain/mod.rs`

arminsabouri committed:
fix(btc_wallet): fix clippy warnings

arminsabouri committed:
fix(executor): clippy warnings

Also ran `cargo fmt` project wide. I try to only include fmt change
for relevant to a sepcific feature/refactor/fix but since this is a
cleanup commit I think its appropirate

arminsabouri commented:
Will open up to review once local testing passes

arminsabouri committed:
feature: run sync controller task using tokio runtime

arminsabouri committed:
refactor(poa): move `process_reciepts()` to crate::utils

arminsabouri committed:
refactor(poa): remove `persist_new_block()`

scottmillner commented on crates/consensus/authority/src/block_fetcher.rs:
```diff
@@ -0,0 +1,211 @@
+
+            let block = new_block.block.block.clone();
+            info!(target: "consensus::authority", ?block, "Recieved new block from peer");
+
```
why was this commented out?

arminsabouri replied:
I was deciding where to put these checks. PoA consenus checks we're kinda spread around all over the place. So I consolidated them all in the latest commit 
[1911a1c](https://github.com/botanix-labs/Macbeth/pull/78/commits/1911a1c08a34f7b20ced52984d1bc56be5e7aeca)

scottmillner commented on crates/consensus/utils.rs:
```diff
@@ -0,0 +1,5 @@
+
```
should this file be deleted?

arminsabouri replied:
Oof yes sorry

arminsabouri committed:
refactor(poa): move block import and sync controller task start to `poa/mod.rs`

arminsabouri committed:
refactor(poa): consolidate poa consensus checks to `consenus/common/validation`

arminsabouri committed:
fix(poa): remove unused utils file

arminsabouri commented:
Per ussual the refactor PR blew up in scope and I ended up doing more than I intended. Let's hop on a call to discuss this PR. Should help the review process

arminsabouri commented on bin/reth/src/poa/mod.rs:
```diff
@@ -549,6 +555,14 @@ impl<Ext: RethCliExt> PoaNodeCommand<Ext> {
             block_production_task.start_task().await;
         }));
 
+        ctx.task_executor.spawn(Box::pin(async move {
```
Should these critical?

scottmillner replied:
seems like they should be right? if any of those tasks die, any reason we'd want to continue running the node?

arminsabouri replied:
Fixed

scottmillner reviewed (approved):
🔥 🔥 🔥

arminsabouri committed:
fix(poa): remove unused `get_block_by_hash` method from lib.rs

Also ran cargo fmt and fix warnings found after rebasing

arminsabouri committed:
ci: add protoc compiler to clippy job

arminsabouri committed:
fix(primitives): remove usless conversions

arminsabouri committed:
fix(poa): get latest block number via client in epoch manager

arminsabouri committed:
fix(poa): insert block via blockchain tree engine before sending FCU message

The engine will look for existing block indecies before making the canon
chain. If that block hash does not exist it will assume there is a gap
and reach out to peers. Since we are creating that block peers will not
have it either ending up in a endless sync.

arminsabouri committed:
refactor(poa): tasks related to consenusus layer should be spawned as critical
